### DISCUSSION TITLE
Reccomended way of dealing with buttons and futures?

### DISCUSSION THREAD

**Discussion by AlbertoNoris** [rrousselGit/riverpod#1907] [Q&A] [answered]
Hi I have a question regarding buttons and futures:
My question is what's the reccomended way for implementing such a flow in Riverpod
1. Tap a button
2. The button calls a future (for example for an http request)
3. The button displays a CurcularProgressIndicator 
4. If the futures completes correctly a new page is pushed
5. If the futures completes with an error a snackbar is shown with the error

I saw that there's the issue #1660 that should resolve my problem but in the meantime is there a way of deling with this with riverpod or I need to do it with vanilla Flutter?
Or maybe there's even a different way of dealing with this type of flows with flutter that I'm not aware?
Thanks for the help,
Alberto 
:)

**Comment by rrousselGit** [comment, marked as answer]
You can conditionally watch a provider

Like:

```dart
if (clicked) {
  ref.watch(someProvider);
}
```

So the "1. tap a button" would basically be doing `setState(() => clicked = true)`

**Reply by AlbertoNoris** [reply]
Thanks for the speedy reply!
Unfortunately I couldn't really grasp what you were really saying but I found this solution, that seems to be doing the job for me (even tho I still would've preferred not using StatefullWidgets).

For future reference:
![image](https://user-images.githubusercontent.com/61833464/202118228-c681b90a-de14-47fc-923d-452de4452d16.png)
The `ListOfHttpCalls.function()` could be replaced with a method of an AsyncNotifier and it should work just fine(mind that in doing so the AsyncData of the provider won't controll the rendering of the button using this method).
